### PR TITLE
WIP implementation of OIDC refresh

### PR DESCRIPTION
**Pull request by maleck13**
- [x] WIP of implementations
- [x] Ran local unit tests 
- [ ] TODO Test with OIDC provider (Keycloak)
- [ ] Additional tests created as required
- [ ] ~Add any needed docs~ (addressed in #180)

@guicassolato Putting up this WIP for initial feedback. Still some work and testing to do but wanted to give opportunity for feedback on implementation.

### CODE REVIEW

File: controllers/auth_config_controller.go [hunk 1]
```diff
@@ -66,6 +66,16 @@ func (r *AuthConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	} else if errors.IsNotFound(err) || !Watched(&authConfig.ObjectMeta, r.LabelSelector) {
 		// could not find the resouce: 404 Not found (resouce must have been deleted)
 		// or the resource misses required labels (i.e. not to be watched by this controller)
+		apiConfig := r.Cache.Get(req.String())
+		if nil != apiConfig {
+			for _, identityConifg := range apiConfig.IdentityConfigs {
+				if v, ok := identityConifg.(common.AuthConfigCleaner); ok {
+					if err := v.Clean(ctx); err != nil {
```
Comment (guicassolato) [inline controllers/auth_config_controller.go:73]: Tricky one, as it should be abstracted at this level, but we know from the implementation that each identity config should have no more than one evaluator and therefore no more than one cleaner. Maybe breaking the loop after successful cleaning once to improve performance?

File: api/v1beta1/auth_config_types.go [hunk 2]
```diff
@@ -143,6 +143,8 @@ type Identity_OidcConfig struct {
 	// Authorino will append to this value the well-known path to the OpenID Connect discovery endpoint (i.e. "/.well-known/openid-configuration"), used to automatically discover the OpenID Connect configuration, whose set of claims is expected to include (among others) the "jkws_uri" claim.
 	// The value must coincide with the value of  the "iss" (issuer) claim of the discovered OpenID Connect configuration.
 	Endpoint string `json:"endpoint"`
+	// Decides how long to wait before refreshing the OIDC configuration
+	TTL int `json:"ttl"`
```
Comment (guicassolato) [inline api/v1beta1/auth_config_types.go:147]: I like "TTL" and no need to change this IMO. Just to leave a comment, for the record, that OPA calls its analogous property `force_cache_duration_seconds`. Ref: https://www.openpolicyagent.org/docs/latest/oauth-oidc/

File: api/v1beta1/auth_config_types.go [hunk 3]
```diff
@@ -143,6 +143,8 @@ type Identity_OidcConfig struct {
 	// Authorino will append to this value the well-known path to the OpenID Connect discovery endpoint (i.e. "/.well-known/openid-configuration"), used to automatically discover the OpenID Connect configuration, whose set of claims is expected to include (among others) the "jkws_uri" claim.
 	// The value must coincide with the value of  the "iss" (issuer) claim of the discovered OpenID Connect configuration.
 	Endpoint string `json:"endpoint"`
+	// Decides how long to wait before refreshing the OIDC configuration
```
Comment (guicassolato) [inline api/v1beta1/auth_config_types.go:146]:
```suggestion
	// Decides how long to wait before refreshing the OIDC configuration (in seconds)
```

File: pkg/config/identity/oidc.go [hunk 4]
```diff
@@ -95,3 +98,38 @@ func (oidc *OIDC) GetURL(name string, ctx context.Context) (*url.URL, error) {
 		return endpoint, nil
 	}
 }
+
+func (oidc *OIDC) configureProviderRefresh(ttl int, ctx context.Context) chan bool {
+	if ttl <= 0 {
+		return nil
+	}
+	done := make(chan bool, 1)
+	duriation := time.Duration(ttl) * time.Second
```
Comment (guicassolato) [inline pkg/config/identity/oidc.go:107]:
```suggestion
	duration := time.Duration(ttl) * time.Second
```

**Review by guicassolato — COMMENTED** [review]
Some minor polishing still pending, but it looks great. Thank you, @maleck13! Really amazing job.

File: controllers/auth_config_controller.go [hunk 5]
```diff
@@ -66,6 +66,16 @@ func (r *AuthConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	} else if errors.IsNotFound(err) || !Watched(&authConfig.ObjectMeta, r.LabelSelector) {
 		// could not find the resouce: 404 Not found (resouce must have been deleted)
 		// or the resource misses required labels (i.e. not to be watched by this controller)
+		apiConfig := r.Cache.Get(req.String())
+		if nil != apiConfig {
```
[the remark this answers is not in the document]
Comment (maleck13) [reply controllers/auth_config_controller.go:70]: No special reason, it is just a habit I have likely picked up from other languages also.

**Comment by guicassolato** [comment]
Superseded by #197

**Pull request closed without merging**